### PR TITLE
Support custom areas

### LINKED ISSUE #83: Support custom areas (in addition to spot areas)
* **Please explain your feature request in a one or two sentences.**
Currently the plug-in only supports spot areas defined upfront in the (Yeedi) app. Those are mapped to so called custom areas in the plug-in. I'd like to be able to also map actual custom areas (defined by coordinates) as custom areas in the plug-in.

* **Is your feature request related to a problem? Please describe.**
We have a combined living and dining room, which I'd like to be able to clean separately. Unfortunately the room was recognized as one room during the map creation (no matter what I tried) and at least the Yeedi app does not offer a feature to edit the map afterwards. Thus I need to work with custom areas (in the Yeedi app) defined by coordinates to do so.

* **Any particular ECOVACS devices that this relates to?**
I'm using a Yeedi Mop Station.

* **Anything else?**
I suggest to integrate the new feature in the device specific configuration along with the support of spot areas. One could provide the coordinates (in my case e.g. for the living room) instead of a spot area ID. The plug-in would need to distinguish these cases and trigger either a spot area or custom area cleaning. See [link](https://github.com/mrbungle64/ecovacs-deebot.js/wiki/CustomArea) for details re: triggering a custom area cleaning using the library at hand.

### PR DESCRIPTION
**Pull request by apfelnutzer**
@bwp91, I'm not done yet, but you can already configure custom areas along with spot areas and trigger custom area cleanings. 

Still to be done:
- refactoring of (old) "custom area" to "predefined area" as a more generic term throughout the code
- readme/howto adjustments
- cleanup of TODOs

Please have a look, if you like!

Closes bwp91/homebridge-deebot#83

### CODE REVIEW

**Comment by apfelnutzer** [comment]
@bwp91, btw - is there any way to migrate the configuration? Because I renamed the "command" to "spotAreaCommand"...

**Comment by apfelnutzer** [comment]
@bwp91, I will externalize the string literal for the new log entries later on, too...

**Comment by bwp91** [comment]
> is there any way to migrate the configuration? Because I renamed the "command" to "spotAreaCommand"...

unfortunately not :)

a name change like this would be a breaking change for all users who use this. i would prefer it not to be if possible.

**Comment by bwp91** [comment]
that said, if you can code to accommodate for both naming conventions, (ie making backward compatible) then this is okay

**Comment by apfelnutzer** [comment]
@bwp91, I think I've found a way to handle this: old commands always have priority over new commands (spot area/custom area). I included a hint in the configuration dialog asking the user to move those entries manually to the new section (as spot areas). At some point in the future (maybe the next big release, when there are needed configuration changes anyway) one could remove this code (see last commit and appropriate comments in the code). 

What do you think?

**Comment by apfelnutzer** [comment]
@bwp91, I'm done so far. Only thing left is to adjust the documentation. I'd suggest the following:

- I'd create a copy the current Custom Area wiki page and reframe and adjust it to Predefined Areas
- I'd then adjust the link from the new configuration page to that new page
- I'd moreover add a hint to the old Custom Area wiki page that it's related to plugin version <= 5.1.0

Would that be fine with you?

**Comment by bwp91** [comment]
Hey @apfelnutzer 
I would just directly edit the wiki right now to how it should be as if this PR were accepted. 
Once you have done this and marked this PR as ready i will give it a look over and merge it in asap

**Comment by apfelnutzer** [comment]
@bwp91, ok - I will do so tomorrow!

**Comment by apfelnutzer** [comment]
@bwp91, just stumbled upon two more things, when updating the wiki... 

First, I can't change the images in the wiki as it's your user content. How can I provide you the new screenshots? 

Second, I renamed the "Custom Area" characteristic to the more generic term "Predefined Area" too, as it now comprises spot and actual custom areas. Maybe that's not an issue because the UUID did not change, but I'm not sure... In the worst case users would have to adjust their scenes, which we would like to avoid. Unfortunately I can't test this without removing and adding the characteristic again, which would break the scenes anyway. Do you know if changing only the name breaks the scenes?

**Comment by apfelnutzer** [comment]
@bwp91: Update re: the images: I managed to upload my images!

**Comment by apfelnutzer** [comment]
@bwp91, wiki is updated. Please note, that the custom area link on the README does not work right now. I've update the link on my branch.

Please review & merge (don't forget to consider my second question above, but I think we're fine)!

**Comment by bwp91** [comment]
For the images I cheat a bit, and go to create an issue on this repo, upload the image into the textbox grab the url and use that in the wiki!

**Comment by apfelnutzer** [comment]
Well I figured out, that you can just drag and drop the images!

File: config.schema.json [hunk 1]
```diff
@@ -126,123 +126,663 @@
...
             },
             "command1": {
-              "title": "Custom Area 1",
+              "title": "Spot Area 1",
               "type": "string",
-              "description": "Enter a 'Spot Area' number or multiple numbers separated by a comma.",
+              "description": "Enter a spot area ID or multiple IDs separated by a comma.",
               "condition": {
-                "functionBody": "return (model?.devices?.[arrayIndices]?.deviceId?.length > 0 && !model?.devices?.[arrayIndices]?.ignoreDevice);"
+                "functionBody": "return (model?.devices?.[arrayIndices]?.deviceId?.length > 0 && !model?.devices?.[arrayIndices]?.ignoreDevice  && model?.devices?.[arrayIndices]?.command1);"
```
Comment (bwp91) [inline config.schema.json:133]: please check this doesn't hide the field if nothing has been entered (eg for new users) or for an existing user that deletes the field)

File: config.schema.json [hunk 2]
```diff
@@ -126,123 +126,663 @@
...
             },
             "command2": {
-              "title": "Custom Area 2",
+              "title": "Spot Area 2",
               "type": "string",
-              "description": "Enter a 'Spot Area' number or multiple numbers separated by a comma.",
+              "description": "Enter a spot area ID or multiple IDs separated by a comma.",
               "condition": {
-                "functionBody": "return (model?.devices?.[arrayIndices]?.deviceId?.length > 0 && !model?.devices?.[arrayIndices]?.ignoreDevice && model?.devices?.[arrayIndices]?.command1?.length > 0);"
+                "functionBody": "return (model?.devices?.[arrayIndices]?.deviceId?.length > 0 && !model?.devices?.[arrayIndices]?.ignoreDevice && model?.devices?.[arrayIndices]?.command2);"
```
Comment (bwp91) [inline config.schema.json:141]: again the check above and for the rest of the commands below

File: lib/utils/custom-chars.js [hunk 3]
```diff
@@ -5,7 +5,7 @@ export default class {
     this.hapChar = api.hap.Characteristic;
     this.uuids = {
       maxSpeed: 'E963F001-079E-48FF-8F27-9C2605A29F52',
-      customArea: 'E963F002-079E-48FF-8F27-9C2605A29F52',
+      predefinedArea: 'E963F002-079E-48FF-8F27-9C2605A29F52',
```
Comment (bwp91) [inline lib/utils/custom-chars.js:8]: as we spoke about, please check to see if this name change is backward compatible

File: package.json [hunk 4]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "homebridge-deebot",
   "alias": "Deebot",
-  "version": "5.1.0",
+  "version": "5.1.0-support-custom-areas",
```
Comment (bwp91) [inline package.json]: if you can remove this change that would be super 👍

Comment (apfelnutzer) [reply package.json]: This was meant just for me to be able to recognize the development version when installed. I thought you would just not merge it... I should probably have kept that local. Removed.

File: config.schema.json (hunk 1 above)
Comment (apfelnutzer) [reply config.schema.json:133]: This is my intention - those fields  named "command1" to "command15" contain the old custom area commands that are now meant to be migrated to the new predefined areas of type spot area. 

I kept those fields for backwards compatibility reasons. The migration needs to be done manually. Until then, everything works as before. I included a hint for the users (see line 864): 

```
Please move those entries to separate areas of type 'Spot Area' below to make use of the new note or custom area feature. IMPORTANT: You need to empty the old commands as they have priority for backwards compatibility reasons.
```

As soon as the user has migrated the old commands and emptied the fields they will disappear. New users are not meant to create the old commands anymore, but directly use the predefined areas instead.

File: config.schema.json (hunk 2 above)
Comment (apfelnutzer) [reply config.schema.json:141]: See above.

File: lib/utils/custom-chars.js (hunk 3 above)
Comment (apfelnutzer) [reply lib/utils/custom-chars.js:8]: I will check this later today!

Comment (apfelnutzer) [reply lib/utils/custom-chars.js:8]: Checking this I found (and fixed) an error re: the condition whether or not to add/remove the predefined area characteristic: so far it was removed when there was not defined at least one predefined area not taking the old custom areas into consideration too...

That being fixed: the renaming does not take effect and thus does not break the configured scenes until the characteristic is removed and added again (which is not the case during a regular update). The scenes still work although the characteristic has been renamed in the background - this is probably due to the fact that the UUID did not change. 

So, the renaming is backwards compatible from my point of view.

**Review by bwp91 — APPROVED** [review]

**Pull request merged**